edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/Metric.cs

```diff
@@ -4,17 +4,17 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Diagnostics
 {
     using System;
     using System.Collections.Generic;
-    using System.Text;
+    using System.Linq;
     using Microsoft.Azure.Devices.Edge.Util;
 
-    public sealed class Metric : IEquatable<Metric>
+    public class Metric : IEquatable<Metric>
     {
         public DateTime TimeGeneratedUtc { get; }
         public string Name { get; }
         public double Value { get; }
-        public string Tags { get; }
+        public IReadOnlyDictionary<string, string> Tags { get; }
 
-        public Metric(DateTime timeGeneratedUtc, string name, double value, string tags)
+        public Metric(DateTime timeGeneratedUtc, string name, double value, IReadOnlyDictionary<string, string> tags)
```

Comment:
Since tags are used to calculate metric key and IReadOnlyDictionary doesn't protect changes to underlying dictionary.  If underlying dictionary is changed, this.Tags.GetHashCode() will be changed as well.  Does this cause any problem as metric key will be different?

Reply:
I think this is a problem; even with same key-value pair in a dictionary, GetHashCode can be different.  You can't use GetHashCode from a dictionary to calculate metric key.

Reply:
https://stackoverflow.com/questions/980390/consistent-hashcodes-for-dictionaries-in-c-sharp

I read this post and it says that for reference types, the expectation is that different references will have different hash codes even if equals() returns true. Thus, for the same reason Dictionary.GetHashCode() returns different hashes for two dicts with the same keys/vals, the Metric should also return different hashes for two Metrics with the same data.

Reply:
Also this method should actually be an override of GetHashCode. The way it is used elsewhere in the code uses a dictionary with this custom function, whereas if we overrode GetHashCode then the dictionary could be a set instead. We can discuss tomorrow.